src/sedpack/io/dataset_iteration.py

```diff
@@ -683,34 +685,131 @@ def as_numpy_iterator_rust(  # pylint: disable=too-many-arguments
 
         def to_dict(example):
             result = {}
             for np_bytes, attribute in zip(
-                    example, self.dataset_structure.saved_data_description):
+                    example, dataset.dataset_structure.saved_data_description):
                 result[attribute.name] = IterateShardFlatBuffer.decode_array(
                     np_bytes=np_bytes,
                     attribute=attribute,
                     batch_size=0,
                 )
             return result
 
-        with _sedpack_rs.RustIter(
+        self._to_dict = to_dict
+
+    def __enter__(self):
+        """Enter the context manager (takes care of freeing memory held by
+        Rust).
+        """
+        return self
+
+    def __exit__(self, exc_type: Optional[Type[BaseException]],
+                 exc_value: Optional[BaseException],
+                 exc_tb: Optional[TracebackType]) -> None:
+        """Drop the rust data structure holding content of open files and
+        future examples.
+        """
+        if self._rust_iter is not None:
+            self._rust_iter.__exit__(exc_type, exc_value, exc_tb)
+
+    def __call__(self) -> Iterable[ExampleT] | Iterable[T]:
+        """Return an iterable.
+        """
+        yield from self._single_iter()
+        while self._repeat:
+            yield from self._single_iter()
+
+    def _single_iter(self) -> Iterable[ExampleT] | Iterable[T]:
+        """Iterate the dataset once.
+        """
+        if self._rust_iter is None:
+            shard_paths: list[str] = list(
+                self._dataset.as_numpy_common(
+                    split=self._split,
+                    shards=self._shards,
+                    shard_filter=self._shard_filter,
+                    repeat=False,
+                    shuffle=self._shuffle,
+                ))
+
+            self._rust_iter = _sedpack_rs.RustIter(
                 files=shard_paths,
-                repeat=repeat,
-                threads=file_parallelism,
-                compression=self.dataset_structure.compression,
-        ) as rust_iter:
-            example_iterator = map(to_dict, iter(rust_iter))
-            if process_record:
-                yield from map(process_record, example_iterator)
-            else:
-                yield from example_iterator
+                repeat=False,
+                threads=self._file_parallelism,
+                compression=self._dataset.dataset_structure.compression,
+            )
+            self._rust_iter.__enter__()  # pylint: disable=unnecessary-dunder-call
+        elif not self._rust_iter.can_iterate:
+            self._rust_iter.__enter__()  # pylint: disable=unnecessary-dunder-call
+
+        example_iterator = map(self._to_dict, iter(self._rust_iter))
+        if self._process_record:
+            yield from map(self._process_record, example_iterator)
+        else:
+            yield from example_iterator
+
+        self._rust_iter.__exit__(None, None, None)
```

Comment:
Or more precisely at the boundary of PyO3.

Opinions welcome :)

Reply:
Let's leave it like this then.
I was just making sure (also with the previous comment) that the code was not possible to rewrite using `with` statements.

Reply:
I am happy for comments and questions. This feels non-trivial to me or I am missing something obvious.

`_sedpack_rs.RustIter` is a Rust data structure which holds data unknown to reference counting (will be dropped at `_sedpack_rs.RustIter.__exit__` call (I could also implement droppable).

`tf.data.Dataset.from_generator` expects a callable which returns an iterable. The problem is that it does call it even without exhausting the previous iterable. I need to prevent leaking data by creating multiple instances of `_sedpack_rs.RustIter`.

Solutions I see:

1. Wrap context manager similar to `RustGenerator` like follows:
```
with RustGenerator(...) as gen:
  tf.data.Dataset.from_generator(gen, ...)
```
The downside is that this is not very handy API (for training I create two datasets -> two indentation levels).
The good part is that we are guaranteed to clean up.

2. Depend on `__del__` being called. With CPython I would be fine. But we would need to depend on TF dropping the reference (it should, I have not tested, but it should) which is not documented anywhere.
The good part is nice API.

3. Changing `RustGenerator.__call__` into:
```
def __call__(self):
  with _sedpack_rs.RustIter() as rust_iter:
    yield from rust_iter
```
This should eventually clean all instances.
The downside is that I am not sure (implementation dependent) what happens when TF calls again (possibly from a different thread or process).